### PR TITLE
feat: add retry / delay on teststep

### PR DESCRIPTION
**Pull request by yesnault**
Signed-off-by: Yvonnick Esnault <yvonnick.esnault@corp.ovh.com>

### CODE REVIEW

File: process.go [hunk 1]
```diff
@@ -233,31 +224,61 @@ func runTestCase(ts *TestSuite, tc *TestCase, l *log.Entry, detailsLevel string)
...
+	l.Infof("end")
+}
+
+func runTestStep(e *executorWrap, tc *TestCase, step TestStep, l *log.Entry, detailsLevel string) {
+
+	var isOK bool
+	var errors []Failure
+	var failures []Failure
+
+	for retry := 0; retry <= e.retry; retry++ {
```
Comment (fsamin) [inline process.go]: for retry := 0; retry < e.retry && !isOK ; retry++ {

File: process.go [hunk 2]
```diff
@@ -233,31 +224,61 @@ func runTestCase(ts *TestSuite, tc *TestCase, l *log.Entry, detailsLevel string)
...
-		if detailsLevel != DetailsLow {
-			bars[ts.Package].Increment()
+			isOK, errors, failures = applyChecks(result, step, nil, l)
 		}
-		if len(tc.Failures) > 0 {
+		if isOK {
 			break
 		}
+		// add errors / failures, only for last retry
+		if retry == e.retry {
```
Comment (fsamin) [inline process.go]: this should not be in the for loop

File: types.go [hunk 3]
```diff
@@ -2,8 +2,51 @@ package venom
...
+const (
+	// DetailsLow prints only summary results
+	DetailsLow = "low"
+	// DetailsMedium prints progress bar and summary
+	DetailsMedium = "medium"
+	// DetailsHigh prints progress bar and details
+	DetailsHigh = "high"
 )
 
+type (
```
Comment (fsamin) [inline types.go]: if you decide to make a `type block`, you should put every in it.
it's up to you, but only one `type block` of repeat `type` everywhere

**Comment by fsamin** [comment]
we need timeout on executorwrapper too :)
and a `context.Context` for cancelation

**Comment by fsamin** [comment]
i could be interesting to be able to user venom to run golang unit tests, by initializing a testcase with a slice of steps and a *testing.T (of *testing.B)

File: process.go (hunk 1 above)
Comment (yesnault) [inline process.go]: it's <=, not <, otherwise, with retry 0 (default value) -> nothing will be runned

File: process.go (hunk 2 above)
Comment (yesnault) [inline process.go]: fixed

File: types.go (hunk 3 above)
Comment (yesnault) [inline types.go]: fixed

File: process.go (hunk 1 above)
Comment (yesnault) [inline process.go]: fixed

**Comment by yesnault** [comment]
@fsamin timeout : yep, but in another PR
@fsamin run golang unit tests --> 
```yaml
steps:
- type: go
  func: TestGetApplicationWithTriggersHandler
  package: github.com/ovh/cds/engine/api
```
--> like that ?

Whatever, it should be in another PR.

**Comment by fsamin** [comment]
@yesnault 
 - step timeout => #4 
 - usability in unit test => #5

**Pull request merged**